test/boost/mutation_reader_test.cc

```diff
@@ -2361,6 +2362,177 @@ SEASTAR_THREAD_TEST_CASE(test_multishard_streaming_reader) {
+            assert(tested_muts.size() == muts_cnt);
+            utils::clear_gently(tested_muts).get();
+
+        }
+        testlog.info("Before flush memtable");
+        env.db().invoke_on_all([] (auto& db) -> future<> {
+            auto s = db.find_column_family("multishard_streaming_reader_ks", "test").schema();
+            auto& t = db.find_column_family(s);
+            return t.flush();
+        }).get();
+        testlog.info("After flush memtable");
+        auto t3 = std::chrono::steady_clock::now();
+        {
+            const auto compaction_time = gc_clock::time_point();
+
+            struct sst_and_shard {
+                unsigned shard;
+                sstring filename;
+            };
+
+            std::vector<foreign_ptr<lw_shared_ptr<utils::chunked_vector<sstables::sstable_files_snapshot>>>> ptrs;
+            ptrs.resize(smp::count);
+
+            auto sstable_filenames = env.db().map_reduce0(
+                [token_range, &ptrs] (auto& db) -> future<std::vector<sst_and_shard>> {
+                    std::vector<sst_and_shard> res;
+                    auto s = db.find_column_family("multishard_streaming_reader_ks", "test").schema();
+                    auto& t = db.find_column_family(s);
+                    auto sstables = co_await t.take_storage_snapshot(token_range);
+                    for (auto& sst : sstables) {
+                        testlog.info("Get sst filename={}", sst.sst->get_filename());
+                        res.push_back(sst_and_shard{this_shard_id(), sst.sst->get_filename()});
+                    }
+                    ptrs[this_shard_id()] = make_foreign(make_lw_shared<utils::chunked_vector<sstables::sstable_files_snapshot>>(std::move(sstables)));
+                    co_return res;
+                },
+                std::vector<sst_and_shard>{},
+                [] (std::vector<sst_and_shard> a, std::vector<sst_and_shard> b) { std::move(b.begin(), b.end(), std::back_inserter(a)); return a;
+            }).get();
+
+            auto s = env.local_db().find_column_family("multishard_streaming_reader_ks", "test").schema();
+            auto& t = env.db().local().find_column_family(s);
+            auto erm = t.get_effective_replication_map();
+            auto& sstm = t.get_sstables_manager();
+            auto sstable_set = make_lw_shared<sstables::sstable_set>(sstables::make_partitioned_sstable_set(s, false));
+
+            size_t nr_local_sstables = 0;
+            size_t nr_remote_sstables = 0;
+            // Load sst from this shard into _sstable_set
+            auto& sstables_on_this_shard = ptrs[this_shard_id()];
+            if (sstables_on_this_shard) {
+                for (auto& sst : *sstables_on_this_shard) {
+                    auto insert = sstable_set->insert(sst.sst);
+                    testlog.debug("Add local sst filename={} insert={}", sst.sst->get_filename(), insert);
+                    ++nr_local_sstables;
+
+                }
+            }
+
```

Comment:
I might be wrong, but this seems a copy of what we have in repair_reader::make_reader(). cannot we extract common parts into functions and reuse them here?

Reply:
We could but the PoC here is about exploring different solutions. We might use one of them in repair_reader.